go-controller/pkg/ovn/controller/admin_network_policy/metrics.go

```diff
@@ -0,0 +1,130 @@
+	reg := prometheus.NewPedanticRegistry()
+	cc := ANPControllerCollector{ANPController: c}
+	prometheus.WrapRegistererWith(prometheus.Labels{"zone": c.zone}, reg).MustRegister(cc)
+	// Add the standard process and Go metrics to the custom registry.
+	reg.MustRegister(
+		// expose process metrics like CPU, Memory, file descriptor usage etc.
+		collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}),
+		// expose Go runtime metrics like GC stats, memory stats etc.
+		collectors.NewGoCollector(),
+	)
+}
+
+// Describe is implemented with DescribeByCollect. That's possible because the
+// Collect method will always return the same two metrics with the same two
+// descriptors.
+func (cc ANPControllerCollector) Describe(ch chan<- *prometheus.Desc) {
+	prometheus.DescribeByCollect(cc, ch)
+}
+
+func (cc ANPControllerCollector) fetchANPRuleCountMetric(anpCache map[string]*adminNetworkPolicyState) map[string]map[string]int {
+	ruleCount := map[string]map[string]int{
+		string(libovsdbutil.ACLIngress): {
+			string(anpapi.AdminNetworkPolicyRuleActionAllow): 0,
+			string(anpapi.AdminNetworkPolicyRuleActionPass):  0,
+			string(anpapi.AdminNetworkPolicyRuleActionDeny):  0,
+		},
+		string(libovsdbutil.ACLEgress): {
+			string(anpapi.AdminNetworkPolicyRuleActionAllow): 0,
+			string(anpapi.AdminNetworkPolicyRuleActionPass):  0,
+			string(anpapi.AdminNetworkPolicyRuleActionDeny):  0,
+		},
+	}
+	for _, state := range anpCache {
+		allow, pass, deny := cc.countRules(state.ingressRules)
+		ruleCount[string(libovsdbutil.ACLIngress)][string(anpapi.AdminNetworkPolicyRuleActionAllow)] += allow
+		ruleCount[string(libovsdbutil.ACLIngress)][string(anpapi.AdminNetworkPolicyRuleActionPass)] += pass
+		ruleCount[string(libovsdbutil.ACLIngress)][string(anpapi.AdminNetworkPolicyRuleActionDeny)] += deny
+		allow, pass, deny = cc.countRules(state.egressRules)
+		ruleCount[string(libovsdbutil.ACLEgress)][string(anpapi.AdminNetworkPolicyRuleActionAllow)] += allow
+		ruleCount[string(libovsdbutil.ACLEgress)][string(anpapi.AdminNetworkPolicyRuleActionPass)] += pass
+		ruleCount[string(libovsdbutil.ACLEgress)][string(anpapi.AdminNetworkPolicyRuleActionDeny)] += deny
+	}
+	return ruleCount
+}
+
+func (cc ANPControllerCollector) countRules(rules []*gressRule) (int, int, int) {
+	var passCount, allowCount, denyCount int
+	for _, rule := range rules {
+		switch rule.action {
+		case nbdb.ACLActionAllowRelated:
+			allowCount++
+		case nbdb.ACLActionDrop:
+			denyCount++
+		case nbdb.ACLActionPass:
+			passCount++
+		default:
+			panic(fmt.Sprintf("Failed to count rule type: unknown acl action %s", rule.action))
+		}
+	}
+	return allowCount, passCount, denyCount
+}
+
+// Collect first triggers the fetchANPRuleCountMetric. Then it
+// creates constant metrics for each host on the fly based on the returned data.
+//
+// Note that Collect could be called concurrently, so we depend on
+// fetchANPRuleCountMetric to be concurrency-safe.
```

Comment:
```suggestion
// fetchANPRuleCountMetric to be concurrency-safe.
```
this was the most concerning bit, I thought collectors can help me because my caches do change dynamically over time, but if this means I need to lock my cache, then might as do what I was doing previously without using collectors because my controller anyways had a central lock on the caches as we counted the metrics.

Reply:
yes you likely want a lock to avoid races.

Reply:
done. did a readlock